File: src/main/java/com/grammr/domain/value/language/Inflection.java
package com.grammr.domain.value.language;

import com.grammr.domain.enums.features.Feature;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.constraints.NotNull;
import java.util.Set;
import lombok.Builder;

/**
 * Represents the inflected form of a word. For instance, "gehst" would have the following features:
 * lemma: gehen
 * inflectedForm: gehst
 * features: Number: Sing, Person: Second, Tense: Present
 */

@Builder
public record Inflection(
    @NotNull
    @Schema(
        description = "Lemma of the inflected word",
        example = "laufen"
    )
    String lemma,
    @NotNull
    @Schema(
        description = "Inflected form of the word",
        example = "läufst"
    )
    String inflected,
    @NotNull
    @Schema(
        description = "Set of features associated with the inflected form",
        example = "[{\"type\": \"NUMBER\", \"value\": \"SING\", \"fullIdentifier\": \"Singular\"}, {\"type\": \"PERSON\", \"value\": \"SECOND\", \"fullIdentifier\": \"Second Person\"}, {\"type\": \"TENSE\", \"value\": \"PRES\", \"fullIdentifier\": \"Present Tense\"}]"
    )
    Set<Feature> features
) {

}

File: sidecars/inflection-ru/inflection/domain/inflection.py
from inflection.domain.feature import Feature
from pydantic import BaseModel


class Inflection(BaseModel):
    lemma: str
    inflected: str
    features: set[Feature]

    def json(self, **kwargs) -> dict:
        return {
            "lemma": self.lemma,
            "inflected": self.inflected,
            "features": [feature.json() for feature in self.features],
        }


class Inflections(BaseModel):
    part_of_speech: str
    lemma: str
    inflections: list[Inflection]

    def json(self, **kwargs) -> dict:
        return {
            "partOfSpeech": self.part_of_speech,
            "lemma": self.lemma,
            "inflections": [inflection.json() for inflection in self.inflections],
        }
